TestValidateNetworkPolicy refactor

danielamlins opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind cleanup
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
Boilerplate NetworkPolicy struct and a helper function that applies tweaks to create specific test cases. 

#### Which issue(s) this PR fixes:

<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Related to umbrella issue #99005 (I don't think we have an specific issue for each test case or test file)

#### Special notes for your reviewer:
To be honest, there is not that much in common between the NetworkPolicies used in each test, so I almost wrote one new tweak per new case. But even so I think that the tweaks helped to make clearer what exactly is changing (especially for the validations that should fail).

#### Does this PR introduce a user-facing change?
NONE
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

danielamlins committed:
Add tweaking functions to TestValidateNetworkPolicy

Signed-off-by: Daniela Lins <danielamlins@gmail.com>

danielamlins committed:
Add tweaks and use them to build error structs

Signed-off-by: Daniela Lins <danielamlins@gmail.com>

danielamlins committed:
Add IPV6 IPBlock to success cases

Signed-off-by: Daniela Lins <danielamlins@gmail.com>

rikatz commented:
/ok-to-test

danielamlins committed:
Fixed gofmt issues

Signed-off-by: Daniela Lins <danielamlins@gmail.com>

danielamlins commented:
@thockin

danielamlins commented:
Seems like a flake. :)
/retest

rikatz commented:
/assign @rikatz
/assign @thockin 

I'll take a look as soon as Tim is happy with the closure approach :)

IMO maybe using a simpler function would be more 'readable' (like https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/networking/networkpolicy/strategy_test.go#L35) but I may be missing something.

Anyway, the approach used in this PR is pretty much similar to the one Tim suggested on old file. I'll try to take a look tomorrow, but if Tim reaches first probably he will give a better direction :)

rikatz commented:
/sig network

thockin commented on pkg/apis/networking/validation/validation_test.go:
```diff
@@ -33,1022 +33,493 @@ import (
+		networkPolicy.Spec.Ingress[0].Ports = []networking.NetworkPolicyPort{
+			{
+				Protocol: nil,
+				Port:     &intstr.IntOrString{Type: intstr.Int, IntVal: 80},
```
I think `intstr.FromInt(80)` should work?

danielamlins replied:
I am not that experienced with golang, and I was having trouble trying to do something like &intstr.fromInt(80), so I needed to declare them as variables on the top and do &port80, for example. I am not sure what is best here.

thockin commented on pkg/apis/networking/validation/validation_test.go:
```diff
@@ -33,1022 +33,493 @@ import (
+		}
+	}
+
+	setIngressFromPodSelector := func(networkPolicy *networking.NetworkPolicy) {
+		networkPolicy.Spec.Ingress[0].From[0].PodSelector = &metav1.LabelSelector{
+			MatchLabels: map[string]string{"c": "d"},
+		}
+	}
+
+	setAlternativeIngressFromPodSelector := func(networkPolicy *networking.NetworkPolicy) {
+		networkPolicy.Spec.Ingress[0].From[0].PodSelector = &metav1.LabelSelector{
+			MatchLabels: map[string]string{"e": "f"},
+		}
+	}
```
In cases like this, does it make more sense to do:

```
setIngressFromPodSelector := func(k, v, string) func(*networking.NetworkPolicy) {
    return func(networkPolicy *networking.NetworkPolicy) {
￼		networkPolicy.Spec.Ingress[0].From[0].PodSelector = &metav1.LabelSelector{
￼			MatchLabels: map[string]string{k: v},
￼		}
￼	}
}
```

Then call it as `setIngressFromPodSelector("c", "d")` or `setIngressFromPodSelector("e", "f")`.  In this way you can sort of build up a mini functional DSL.

Even for cases where you don't have multiple identical-but-for-args implementations, this can make the test cases clearer - the goal is to make each test-case as obvious and terse as possible, and not terser.

thockin commented on pkg/apis/networking/validation/validation_test.go:
```diff
@@ -33,1022 +33,493 @@ import (
+	}
+
+	setEmptyExcept := func(networkPolicy *networking.NetworkPolicy) {
+		networkPolicy.Spec.Ingress[0].From[0].IPBlock.Except = []string{"", " "}
```
This seems to be testing 2 cases, but we don't really verify that BOTH fail?

thockin commented on pkg/apis/networking/validation/validation_test.go:
```diff
@@ -33,1022 +33,493 @@ import (
+	}
+
+	errorCases := map[string]*networking.NetworkPolicy{
+		"namespaceSelector and ipBlock":                     makeNetworkPolicyCustom(setIngressEmptyFirstElement, setIngressFromEmptyFirstElement, setIngressFromNamespaceSelector, setIngressFromIPBlock),
```
This certainly shortens the test, and I think it's MOSTLY clearer.  It's also OK to mix the modes when they are very localized.

e.g.

```
"invalid ingress.ports.protocol":  makeNetworkPolicyCustom(func(networkPolicy *networking.NetworkPolicy) {
	networkPolicy.Spec.Ingress = makeEmptyIngress()
	networkPolicy.Spec.Ingress[0].Ports = []networking.NetworkPolicyPort{{
		Protocol: &protocolICMP,
		Port:     &intstr.IntOrString{Type: intstr.Int, IntVal: 80},
	}}
})
```

or even

```
e.g.

```
"invalid ingress.ports.protocol":  makeNetworkPolicyCustom(func(networkPolicy *networking.NetworkPolicy) {
	networkPolicy.Spec.Ingress = makeValidIngress()
	networkPolicy.Spec.Ingress[0].Ports[0].Protocol = &protocolICMP
```

The goal being that the reader of the testcase can see quickly "what is special
about this case".

thockin reviewed (commented):
Overall I think this is a win

danielamlins committed:
Went through the review notes

- Adapt tweaks to be clearer
- Use intstr.fromInt and intstr.fromStr
- Added more tests to invalid ports

Signed-off-by: Daniela Lins <danielamlins@gmail.com>

danielamlins commented:
Thanks for the review notes thockin and rikatz!  :smile: 
I think I went through all the suggested changes, happy to change anything else.

danielamlins commented:
/retest

rikatz commented:
/sig cli

rikatz commented:
/remove-sig cli
/release-note-none
/retest
Yeah, needed to hack to remove the do-not-merge label :)

rikatz commented:
/triage accepted
/priority backlog

danielamlins commented:
/retest

JornShen commented on pkg/apis/networking/validation/validation_test.go:
```diff
@@ -33,1022 +33,422 @@ import (
+
+	for k, v := range successCases {
+		if errs := ValidateNetworkPolicy(v); len(errs) != 0 {
+			t.Errorf("Expected success for %d, got %v", k, errs)
```
here k is right?  used v?  
I think line 279  can be
```
for _, v := range successCases {
```
k is just a index num.

danielamlins replied:
I tried printing the test case object but I don't think it was clear (it was mostly noise). However, I changed the error message a bit to specify which test case did not pass and included inline comments to make it easier to find the failing test. On the other tests (the ones that look for fail cases), they can print the "name" of the test, but here we can't.

JornShen commented on pkg/apis/networking/validation/validation_test.go:
```diff
@@ -33,1022 +33,422 @@ import (
 	for testName, networkPolicy := range errorCases {
-		if errs := ValidateNetworkPolicy(&networkPolicy); len(errs) == 0 {
+		if errs := ValidateNetworkPolicy(networkPolicy); len(errs) == 0 {
 			t.Errorf("Expected failure for test: %s", testName)
```
I think we can add networkPolicy info in log. If faild, we can check it quickly.

danielamlins replied:
The networkPolicy struct has too many fields and again I think this would be a lot of noise in the test result. Would you print the whole struct or some specific fields?

JornShen reviewed (commented):
two small advise.
:)

danielamlins committed:
Change error msg and comments to easily find test

Signed-off-by: Daniela Lins <danielamlins@gmail.com>

danielamlins commented:
/retest

thockin commented on pkg/apis/networking/validation/validation_test.go:
```diff
@@ -33,1022 +33,439 @@ import (
-					},
+			}
+		}),
+		"invalid ingress.ports.protocol": makeNetworkPolicyCustom(setIngressEmptyPorts, func(networkPolicy *networking.NetworkPolicy) {
```
Another way to spin this sort of case:

```
"invalid ingress.ports.protocol": makeNetworkPolicyCustom(func(networkPolicy *networking.NetworkPolicy) {
	networkPolicy.Spec.Ingress = makeIngress(Port{Protocol: &ICMP})
￼}),
```
```

thockin reviewed (commented):
There are a lot of cases covered here - I am assuming you are mostly just transliterating them.

All of my feedback is around optimizing for read.  This already looks good, but I'm not above raising the bar.  Tell me what you think.

danielamlins committed:
Refactored setIngressPorts and setEgressPorts

Signed-off-by: Daniela Lins <danielamlins@gmail.com>

danielamlins committed:
Make tweaks only reset slices if empty

Signed-off-by: Daniela Lins <danielamlins@gmail.com>

danielamlins committed:
Remove success case comments

Signed-off-by: Daniela Lins <danielamlins@gmail.com>

danielamlins commented:
Hey @thockin 
I went to the review notes and I believe that I  refactored the code accordingly. 
Thanks a lot for the attention given to this PR. I'll save this PR as a reference because of some really nice suggestion that I can use in other projects :smiley: 
Please, let me know if there are any other changes needed.

danielamlins commented:
/retest

danielamlins committed:
Fix empty setter logic

Signed-off-by: Daniela Lins <danielamlins@gmail.com>

danielamlins commented:
/retest

danielamlins commented:
/retest

danielamlins committed:
Delete not used tweak

Signed-off-by: Daniela Lins <danielamlins@gmail.com>

thockin commented on pkg/apis/networking/validation/validation_test.go:
```diff
@@ -33,1022 +33,315 @@ import (
+	}
+
+	setIngressFromEmptyFirstElement := func(networkPolicy *networking.NetworkPolicy) {
+		if networkPolicy.Spec.Ingress == nil {
```
do you even want to test this?  For example: is it useful for a test to set `spec.ingress[]` to a 2-element list and then call this?  I feel like the more opinionated and less subtle these are, the easier they are to comprehend.

thockin replied:
I see the IfEmpty() forms below - I won't hold your feet to the fire on this aspect.  Mostly just something to keep in mind.  The more `if` clauses the tweaks grow, the more likely we are to screw it up :)

thockin commented on pkg/apis/networking/validation/validation_test.go:
```diff
@@ -33,1022 +33,315 @@ import (
+	setIngressFromPodSelector := func(k, v string) func(*networking.NetworkPolicy) {
+		return func(networkPolicy *networking.NetworkPolicy) {
+			setIngressFromIfEmpty(networkPolicy)
+			networkPolicy.Spec.Ingress = []networking.NetworkPolicyIngressRule{
```
super minor, but I think it's nice to eat less vertical space:

```
networkPolicy.Spec.Ingress = []networking.NetworkPolicyIngressRule{{
	From: []networking.NetworkPolicyPeer{{
		PodSelector: &metav1.LabelSelector{
			MatchLabels: map[string]string{k: v},
		},
	}},
}}
```

thockin commented on pkg/apis/networking/validation/validation_test.go:
```diff
@@ -33,1022 +33,315 @@ import (
+	}
+
+	setIngressFromNamespaceSelector := func(networkPolicy *networking.NetworkPolicy) {
+		setIngressFromIfEmpty(networkPolicy)
```
This one is so much shorter than the equialent `FromPodSelector`?  I'd expect them to be congruent

thockin commented on pkg/apis/networking/validation/validation_test.go:
```diff
@@ -33,1022 +33,315 @@ import (
+		}
+	}
+
+	setIngressFromIPBlock := func(networkPolicy *networking.NetworkPolicy) {
```
If you want to call out v4 vs v6, I'd name them both that way `setIngressFromIPBlockIPv4`

thockin commented on pkg/apis/networking/validation/validation_test.go:
```diff
@@ -33,1022 +33,315 @@ import (
+	successCases := []*networking.NetworkPolicy{
+		makeNetworkPolicyCustom(setIngressEmptyFirstElement),
+		makeNetworkPolicyCustom(setIngressFromEmptyFirstElement, setIngressEmptyPorts),
+		makeNetworkPolicyCustom(setIngressPorts(makePort(nil, intstr.FromInt(80), 0), makePort(&protocolTCP, intstr.FromInt(0), 0), makePort(&protocolTCP, intstr.FromInt(443), 0), makePort(&protocolUDP, intstr.FromString("dns"), 0), makePort(&protocolSCTP, intstr.FromInt(7777), 0))),
```
suggest a linewrap for each makePort

thockin commented on pkg/apis/networking/validation/validation_test.go:
```diff
@@ -33,1022 +33,315 @@ import (
+		makeNetworkPolicyCustom(setIngressFromIPBlock),
+		makeNetworkPolicyCustom(setEgressToIPBlock, setPolicyTypesEgress),
+		makeNetworkPolicyCustom(setEgressToIPBlock, setPolicyTypesIngressEgress),
+		makeNetworkPolicyCustom(setEgressPorts(makePort(nil, intstr.FromInt(80), 0), makePort(&protocolTCP, intstr.FromInt(0), 0), makePort(&protocolTCP, intstr.FromInt(443), 0), makePort(&protocolUDP, intstr.FromString("dns"), 0), makePort(&protocolSCTP, intstr.FromInt(7777), 0))),
```
linewrap suggested

thockin commented on pkg/apis/networking/validation/validation_test.go:
```diff
@@ -33,1022 +33,315 @@ import (
+		"namespaceSelector and ipBlock": makeNetworkPolicyCustom(setIngressFromNamespaceSelector, setIngressFromIPBlock),
+		"podSelector and ipBlock":       makeNetworkPolicyCustom(setEgressToPodSelector, setEgressToIPBlock),
+		"missing from and to type":      makeNetworkPolicyCustom(setIngressFromEmptyFirstElement, setEgressToEmptyFirstElement),
+		"invalid spec.podSelector": makeNetworkPolicyCustom(setIngressFromNamespaceSelector, func(networkPolicy *networking.NetworkPolicy) {
```
suggest linewrap on func

thockin reviewed (commented):
Other than mostly formatting stuff (linewraps, etc) this is great.  It's WAY more understandable I think.

danielamlins committed:
Handle int and string port in makePort

Signed-off-by: Daniela Lins <danielamlins@gmail.com>

danielamlins committed:
Reduce vertical space

Signed-off-by: Daniela Lins <danielamlins@gmail.com>

danielamlins committed:
Standardize Namespace and Podselector tweaks

Signed-off-by: Daniela Lins <danielamlins@gmail.com>

danielamlins committed:
Specify IPV4 on tweak names

Signed-off-by: Daniela Lins <danielamlins@gmail.com>

danielamlins committed:
Linewrap for makePort calls and lint fixes

Signed-off-by: Daniela Lins <danielamlins@gmail.com>

thockin commented on pkg/apis/networking/validation/validation_test.go:
```diff
@@ -33,1022 +33,348 @@ import (
+		"invalid policyTypes": makeNetworkPolicyCustom(setEgressToIPBlockIPV4, func(networkPolicy *networking.NetworkPolicy) {
+			networkPolicy.Spec.PolicyTypes = []networking.PolicyType{"foo", "bar"}
+		}),
+		"too many policyTypes": makeNetworkPolicyCustom(setEgressToIPBlockIPV4, func(networkPolicy *networking.NetworkPolicy) {
```
You could define a `setPolicies(p ...string)` tweak

thockin commented on pkg/apis/networking/validation/validation_test.go:
```diff
@@ -33,1022 +33,348 @@ import (
+				Except: []string{"192.168.0.0/24"},
+			}
+		}),
+		"except IPv6 is outside of CIDR range": makeNetworkPolicyCustom(setIngressFromEmptyFirstElement, func(networkPolicy *networking.NetworkPolicy) {
```
If you wanted to get really fancy, maybe you can do something like:

```
makeNetworkPolicyCustom(setIngress(setFrom(setIPBlock("fd00:192:168:1::/64", "fd00:192:168:2::/64""))))`
```

It would probably mean you could remove some tweaks.

This could be a followup, of course ;)

thockin reviewed (commented):
Thanks!

/lgtm
/approve
